README.md
## isdparser 
[![CI](https://github.com/bsnacks000/isdparser/actions/workflows/CI.yaml/badge.svg)](https://github.com/bsnacks000/isdparser/actions/workflows/CI.yaml)


A utility package to help parse noaa isd files from `ftp://ftp.ncdc.noaa.gov/pub/data/noaa`

Turns this:
```
0130010230999992020010100004+69067+018533FM-12+007999999V0200501N001019999999N999999999-00291-00381099661
```

Into this:
```python
{'datestamp': datetime.datetime(2020, 1, 1, 0, 0, tzinfo=datetime.timezone.utc),
  'identifier': '010230-99999',
  'sections': [{'measures': [{'usaf': '010230'},
                             {'wban': '99999'},
                             {'date': '20200101'},
                             {'time': '0000'},
                             {'description': 'USAF SURFACE HOURLY observation',
                              'measure': 'data_source_flag',
                              'value': '4'},
                             {'measure': 'latitude',
                              'unit': 'angular_degrees',
                              'value': 69.067},
                             {'measure': 'longitude',
                              'unit': 'angular_degrees',
                              'value': 18.533},
                             {'description': 'SYNOP Report of surface '
                                             'observation form a fixed land '
                                             'station',
                              'measure': 'code',
                              'value': 'FM-12'},
                             {'measure': 'elevation_dimension',
                              'unit': 'meters',
                              'value': 79.0},
                             {'call_letter_identifier': None},
                             {'description': 'Automated Quality Control',
                              'measure': 'quality_control_process_name',
                              'value': 'V020'}],
                'name': 'control'},
               {'measures': [{'measure': 'wind_observation_direction_angle',
                              'unit': 'angular_degrees',
                              'value': 50.0},
                             {'description': 'Passed all quality control '
                                             'checks',
                              'measure': 'wind_observation_direction_quality_code',
                              'value': '1'},
                             {'description': 'Normal',
                              'measure': 'wind_observation_type_code',
                              'value': 'N'},
                             {'measure': 'wind_observation_speed_rate',
                              'unit': 'meters_per_second',
                              'value': 1.0},
                             {'description': 'Passed all quality control '
                                             'checks',
                              'measure': 'wind_observation_speed_quality_code',
                              'value': '1'},
                             {'measure': 'sky_condition_observation_ceiling_height_dimension',
                              'unit': 'meters',
                              'value': None},
                             {'description': 'Passed gross limits check if '
                                             'element is present',
                              'measure': 'sky_condition_observation_ceiling_quality_code',
                              'value': '9'},
                             {'description': 'Missing',
                              'measure': 'sky_condition_observation_ceiling_determination_code',
                              'value': '9'},
                             {'description': 'No',
                              'measure': 'sky_condition_observation_cavok_code',
                              'value': 'N'},
                             {'measure': 'visibility_observation_distance_dimension',
                              'unit': 'meters',
                              'value': None},
                             {'description': 'Passed gross limits check if '
                                             'element is present',
                              'measure': 'visibility_observation_distance_quality_code',
                              'value': '9'},
                             {'description': 'Missing',
                              'measure': 'visibility_observation_variability_code',
                              'value': '9'},
                             {'description': 'Passed gross limits check if '
                                             'element is present',
                              'measure': 'visibility_observation_quality_variability_code',
                              'value': '9'},
                             {'measure': 'air_temperature_observation_air_temperature',
                              'unit': 'degrees_celsius',
                              'value': -2.9},
                             {'description': 'Passed all quality control '
                                             'checks',
                              'measure': 'air_temperature_observation_air_temperature_quality_code',
                              'value': '1'},
                             {'measure': 'air_temperature_observation_dew_point_temperature',
                              'unit': 'degrees_celsius',
                              'value': -3.8},
                             {'description': 'Passed all quality control '
                                             'checks',
                              'measure': 'air_temperature_observation_dew_point_quality_code',
                              'value': '1'},
                             {'measure': 'atmospheric_pressure_observation_sea_level_pressure',
                              'unit': 'hectopascals',
                              'value': 996.6},
                             {'description': 'Passed all quality control '
                                             'checks',
                              'measure': 'atmospheric_pressure_observation_sea_level_pressure_quality_code',
                              'value': '1'}],
                'name': 'mandatory'}]},
```


### Notes 

The current version parses only the `control` and `mandatory` data sections of the isd record-string. The `additional` data section is extremely inconsistent between records and would require alot of work to properly map in a sane way.

The above schema was constructed with mongo in mind so feel free to fork and modify to your needs. Any changes or additions to the above schema will incur a minor version bump (see below for install).

Missing data for numerical measures are represented as `None` in the python schema. This should map to a database better then symbols like `-7777` or `+999999` which are used in the strings. 

All numerical data is also "scaled" and appropriately signed by the provided `scaling_factor` from the pdf.

I've included a `data-dictionary.md` and a copy of the isd documentation I used to write the parsers in the `extras` folder.

### Install 

For now install from here with pip. The master branch will contain the most up to date version.   
```
pip install git+https://github.com/bsnacks000/isdparser.git
```
or a specific version 
```
pip install git+https://github.com/bsnacks000/isdparser.git@v0.1.0
```

### Usage 


If you want to parse an isd with the default configuration which includes the mappings I created and all points from the `control` and `data` sections, then it is very simple. 

```python
from isdparser import ISDRecordFactory 

# assumes you downloaded a file but could also be a direct to ftp connection.
with open('010230-99999-2020.txt', 'r') as f:
    lines = f.readlines()

# create a list of record objects
records = [ISDRecordFactory().create(line) for line in lines]    

# build the schema 
schema = [r.schema() for r in records]

# do things...
```
This will produce the schema documented above. It can can serialized directly to json, offloaded to mongo or easily flattened and uploaded to a sql database. 

One caveat for using the high level factory is that it will make a root level key by looking up certain values in the control section. This is a convenience and assumes you want to have some easy to manage data integrity based on the usaf, wban and datestamp. These three fields will need to be given in the control section. If for some reason you don't want this behavior I recommend not using the high level factory and creating your own high level schema.

If you want to modify the API to parse less data then simply create new lists of `control` and `mandatory` measures and give these along with names to the `ISDRecordFactory`. You can do this either using a callable or explicitly with lists of Measures. 

HISTORY.md
## History 
------------


* __v0.1.0 (2021-02-15)__
    * initial release

tests/smoke.py


def test_imports():
    import isdparser 


tests/conftest.py
import pytest


@pytest.fixture
def isd_record_string(): 
    return """0130010230999992020010100004+69067+018533FM-12+007999999V0200501N001019999999N999999999-00291-00381099661
    ADDAA199999919MA1999999098701MD1810211+9999OD139900601999REMSYN06401023 16/// /0501 11029 21038 39870 49966 58021 6//// 333 91106="""




tests/test_record.py
import pytest 
import datetime 
import unittest 
tc = unittest.TestCase()

from isdparser.measures import Measure, Section
from isdparser.record import ISDRecord, ISDRecordFactory

def test_isdrecord(mocker): 

    mocker.patch.object(Section, 'schema', return_value={
        'data': 'goes here'
    })

    mocker.patch.object(Measure, 'schema', return_value={
        'more': 'data goes here'
    })

    sections = [Section('', [Measure('',None) for _ in range(3)]) for _ in range(3)]

    # mocker.patch.object(NumericMeasure, 'schema', 
    #     return_value={'measure': 'my_measure', 'value': 42, 'unit': 'my_unit'})


    d = datetime.datetime(2021,1,1,0,0)
    
    with pytest.raises(ValueError):
        ISDRecord(d, 'my-id', [])
    
    record = ISDRecord(d, 'my-id', sections)
    expected = {
        'datestamp': d, 
        'identifier': 'my-id', 
        'sections': [s.schema() for s in sections]
    }

    tc.assertDictEqual(expected, record.schema())


def test_isdrecordfactory_static_methods():

    fac = ISDRecordFactory((),())
    assert fac.parse_line('abcde', 3, 4) == 'd'

    assert fac.make_station_identifier('usaf', 'wban') == 'usaf-wban'

    assert fac.make_datestamp('20210101', '0115'), datetime.datetime(2021,1,1,1,15)



def test_isdrecordfactory_on_record_string(isd_record_string):
    
    result = ISDRecordFactory().create(isd_record_string)

    expected = {
        'datestamp': datetime.datetime(2020, 1, 1, 0, 0, tzinfo=datetime.timezone.utc), 
        'identifier': '010230-99999', 
        'sections': [
            {
                'name': 'control', 
                'measures': [
                    {'measure': 'usaf', 'value':'010230'}, 
                    {'measure': 'wban', 'value': '99999'}, 
                    {'measure': 'date', 'value': '20200101'}, 
                    {'measure': 'time', 'value': '0000'}, 
                    {'measure': 'data_source_flag', 'value': '4', 'description': 'USAF SURFACE HOURLY observation'}, 
                    {'measure': 'latitude', 'value': 69.067, 'unit': 'angular_degrees'}, 
                    {'measure': 'longitude', 'value': 18.533, 'unit': 'angular_degrees'}, 
                    {'measure': 'code', 'value': 'FM-12', 'description': 'SYNOP Report of surface observation form a fixed land station'}, 
                    {'measure': 'elevation_dimension', 'value': 79.0, 'unit': 'meters'}, 
                    {'measure': 'call_letter_identifier', 'value': None}, 
                    {'measure': 'quality_control_process_name', 'value': 'V020', 'description': 'Automated Quality Control'}]}, 
            {
                'name': 'mandatory', 
                'measures': [
                    {'measure': 'wind_observation_direction_angle', 'value': 50.0, 'unit': 'angular_degrees'}, 
                    {'measure': 'wind_observation_direction_quality_code', 'value': '1', 'description': 'Passed all quality control checks'}, 
                    {'measure': 'wind_observation_type_code', 'value': 'N', 'description': 'Normal'}, 
                    {'measure': 'wind_observation_speed_rate', 'value': 1.0, 'unit': 'meters_per_second'}, 
                    {'measure': 'wind_observation_speed_quality_code', 'value': '1', 'description': 'Passed all quality control checks'}, 
                    {'measure': 'sky_condition_observation_ceiling_height_dimension', 'value': None, 'unit': 'meters'}, 
                    {'measure': 'sky_condition_observation_ceiling_quality_code', 'value': '9', 'description': 'Passed gross limits check if element is present'}, 
                    {'measure': 'sky_condition_observation_ceiling_determination_code', 'value': None, 'description': 'Missing'}, 
                    {'measure': 'sky_condition_observation_cavok_code', 'value': 'N', 'description': 'No'}, 
                    {'measure': 'visibility_observation_distance_dimension', 'value': None, 'unit': 'meters'}, 
                    {'measure': 'visibility_observation_distance_quality_code', 'value': '9', 'description': 'Passed gross limits check if element is present'}, 
                    {'measure': 'visibility_observation_variability_code', 'value': None, 'description': 'Missing'}, 
                    {'measure': 'visibility_observation_quality_variability_code', 'value': '9', 'description': 'Passed gross limits check if element is present'}, 
                    {'measure': 'air_temperature_observation_air_temperature', 'value': -2.9, 'unit': 'degrees_celsius'},
                    {'measure': 'air_temperature_observation_air_temperature_quality_code', 'value': '1', 'description': 'Passed all quality control checks'}, 
                    {'measure': 'air_temperature_observation_dew_point_temperature', 'value': -3.8, 'unit': 'degrees_celsius'}, 
                    {'measure': 'air_temperature_observation_dew_point_quality_code', 'value': '1', 'description': 'Passed all quality control checks'}, 
                    {'measure': 'atmospheric_pressure_observation_sea_level_pressure', 'value': 996.6, 'unit': 'hectopascals'}, 
                    {'measure': 'atmospheric_pressure_observation_sea_level_pressure_quality_code', 'value': '1', 'description': 'Passed all quality control checks'}
                ]
            }
        ]
    }

    tc.assertDictEqual(expected, result.schema())

isdparser/record.py
"""Module for creating the parent schema for a single isd record.
"""
    
import datetime
from typing import Any, Dict, List, Tuple, Union

from .measures import Measure, Section, control_measures, mandatory_measures


class ISDRecord(object):
    __slots__ = ('_datestamp', '_identifier', '_sections')
    
    def __init__(self, 
        datestamp: datetime.datetime, 
        identifier: str, 
        sections: List[Section]) -> None:
        """A container for an isd record. This is a top level container suitable for storage with a 
        top level key and appropriate sections. Technically to create a record with this API all that is needed
        is a single control section with date, time usaf and wban.

        Args:
            datestamp (datetime.datetime): A datetime combined from `date` and `time` control section data.
            identifier (str): An identifier of `usaf`-`wban` extracted from control section data.
            sections (List[Section]): A list of section objects. 
        Raises:
            ValueError: If an empty list is passed for sections.
        """
        self._datestamp = datestamp 
        self._identifier = identifier 
        if len(sections) == 0:
            raise ValueError("Must provide at least a single section.")
        self._sections = sections 
        
        
    def schema(self) -> Dict[str, Union[datetime.datetime, str, Dict[str, Any]]]:
        """Return the record schema

        Returns:
            Dict[str, Union[datetime.datetime, str, Dict[str, Any]]]: Dictionary of complete schema.
        """
        return {  # note this is specifically for mongo
            'datestamp': self._datestamp, 
            'identifier': self._identifier, 
            'sections': [s.schema() for s in self._sections]
        } 


class ISDRecordFactory(object):
    __slots__ = ('_control', '_mandatory')
    
    def __init__(self, 
                 control: Tuple[str, List[Measure]]=('control', control_measures()), 
                 mandatory: Tuple[str, List[Measure]]=('mandatory', mandatory_measures())):
        """Creates an ISDRecord from a single line from an isd file. 
        Object configuration uses Section and Measure object APIs internally. 

        Args:
            control (Tuple[str, List[Measure]], optional): A name and list of control measures. 
                Defaults to ('control', control_measures()).
            mandatory (Tuple[str, List[Measure]], optional): A name and list of section measures. 
                Defaults to ('mandatory', mandatory_measures()).
        """
        self._control = control
        self._mandatory = mandatory
        

    @staticmethod 
    def parse_line(line: str, start: int, end: int) -> str:
        """Return a slice of string.

        Args:
            line (str): The line to slice from
            start (int): start slice
            end (int): end slice

        Returns:
            str: The string slice.
        """
        return line[start:end]
    
    
    @staticmethod
    def make_datestamp(yyyymmdd: str, hhmm: str) -> datetime.datetime:
        """Make a datestamp from a set of two strings.

        Args:
            yyyymmdd (str): A string formatted 'YYYYmmdd'
            hhmm (str): A string formatted 'HHMM'

        Returns:
            datetime.datetime: A utc encoded datetime object.
        """
        year, month, day = int(yyyymmdd[:4]), int(yyyymmdd[4:6]), int(yyyymmdd[6:])
        hours, minutes = int(hhmm[:2]), int(hhmm[2:])
        return datetime.datetime(year, month, day, hours, minutes, tzinfo=datetime.timezone.utc)


    @staticmethod
    def make_station_identifier(usaf: str, wban: str) -> str:
        """Make the station identifier string

        Args:
            usaf (str): usaf string
            wban (str): wban string

        Returns:
            str: combined identifier `usaf`-`wban`
        """
        return usaf + '-' + wban
    
    
    def _extract(self, section, name):
        for m in section.measures():
            if name == m.name:
                return m._value 


    def _create_key(self, section, key=()):
        a,b = self._extract(section, key[0]), self._extract(section, key[1])
        assert all((a, b)), f'`{key[0]}` and `{key[1]}` must be included in the control section in order to create a ParentRecord'
        return a, b 


    def create(self, line: str) -> ISDRecord:
        """Create an ISDRecord with a line of isd data extracted from a file.

        Args:
            line (str): A single line of raw isd data.

        Returns:
            ISDRecord: The resulting ISDRecord object.
        """
        control, mandatory = Section(*self._control), Section(*self._mandatory)
        [m.set_value(self.parse_line(line, m.start, m.end)) 
                        for s in (control, mandatory) 
                            for m in s.measures()]
        
        date, time = self._create_key(control, ('date', 'time'))
        usaf, wban = self._create_key(control, ('usaf', 'wban'))

        datestamp = self.make_datestamp(date, time)
        identifier = self.make_station_identifier(usaf, wban)
        
        return ISDRecord(datestamp, identifier, [control, mandatory])

isdparser/measures.py

from . import mappings 
from typing import Any, Callable, Dict, Generator, List, Optional, Tuple, Union


class Position(object): 
    __slots__ = ('start', 'end')
    
    def __init__(self, start: int, end: int) -> None:
        """Container object for the position in the isd file. This should 
        be the exact position as found in the documentation.

        Args:
            start (int): start pos
            end (int): end pos
        Raises:
            ValueError: If start date is greater then end.
        """
        if start > end: 
            raise ValueError("`start` cannot be `greater` then end.")
        self.start = start
        self.end = end


class Measure(object):
    __slots__ = ('_name', '_position', '_missing', '_value')
    
    def __init__(self, name: str, position: Position, missing: Optional[str]=None) -> None:
        """A basic measure found in the isd string, represented as 

        Args:
            name (str): The name of the measure
            position (Position): A position object.
            missing (str, optional): If provided, the value represented as missing.
                Often something like `9999`. Defaults to None.
        """
        self._name = name 
        self._position = position # easier to map  
        self._missing = missing 
        self._value = None
        
    @property 
    def name(self) -> str:
        """Get the name of the Measure

        Returns:
            str: The name of the Measure
        """
        return self._name
    
    @property 
    def start(self) -> int:
        """Retrieve the start from the underlying position. Subtracts 1 from this
        value to give the python list position for a slice.

        Returns:
            int: Provided start position - 1
        """
        return self._position.start - 1 

    @property 
    def end(self) -> int:
        """Retrieve end from the underlying position.

        Returns:
            int: [description]
        """
        return self._position.end
    
    def schema(self) -> Dict[str, Any]:
        """Return the schema as a dictionary mapped name : value

        Returns:
            Dict[str, Any]: A dictionary mapped name value.
        """
        if self._missing is not None and self._value == self._missing:
            val = None
        else:
            val = self._value
        return { 'measure': self._name, 'value': val }
    

    def set_value(self, value: str) -> "Measure":
        """Set the value of the string on the Measure instance returning self.

        Returns:
            Measure: the instance.
        """
        self._value = value 
        return self


class NumericMeasure(Measure):
    
    __slots__ = ('_name', '_position', '_missing', '_value', '_scaling_factor', '_unit')
    
    def __init__(self, 
        name: str, 
        position: Position, 
        missing: Optional[str]=None, 
        scaling_factor: int=1, 
        unit: str='') -> None:
        """Describes a Numeric measure from the isd documentation

        Args:
            name (str): The name of the measure
            position (Position): Measure position
            missing (Optional[str], optional): The repr of a missing value if provided. Defaults to None.
            scaling_factor (int, optional): The scaling factor used to create a float. Defaults to 1.
            unit (str, optional): The unit of measure. Defaults to ''.
        """
        super().__init__(name, position, missing)
        self._scaling_factor = scaling_factor
        self._unit = unit
        
    def schema(self) -> Dict[str, Any]:
        """Return the schema as a dictionary, calculating any values from the string value.

        Returns:
            Dict[str, Any]: Resulting schema mapped measure: name, value: val and unit: unit
        """
        if self._missing is not None and self._value == self._missing: 
            val = None 
        else:
            val = float(self._value) / self._scaling_factor 
        return {
            'measure': self._name, 
            'value': val,
            'unit': self._unit
        }
    

class CategoricalMeasure(Measure): 
    __slots__ = ('_name', '_position', '_missing', '_value', '_mapping')
    
    def __init__(self, 
        name: str, 
        position: Position, 
        mapping: Dict[str, str], 
        missing: Optional[str]=None) -> None:
        """ A CategoricalMeasure. A dictionary mapping values to descriptions must be provided based on the isd docs.

        Args:
            name (str): The name of the measure.
            position (Position): The position of the measure.
            mapping (Dict[str, str]): A dict mapping the encoded value to a description.
            missing (Optional[str], optional): The repr of a missing value if provided. Defaults to None.. Defaults to None.
        """
        super().__init__(name, position, missing)
        self._mapping = mapping 
    
    def schema(self) -> Dict[str, Any]:
        """ Returns the schema as a dictionary. A description maps the repr value to text.

        Returns:
            Dict[str, Any]: Resulting schema mapped meaure : name, value : val, description : text
        """
        if self._missing is not None and self._value in self._missing:
            val = None
        else:
            val = self._value
        return {
            'measure': self._name,
            'value': val,
            'description': self._mapping[self._value]
        }


class Section(object): 
    __slots__ = ('_name', '_measures')
    
    def __init__(self, 
        name: str, 
        measures: Union[Callable[...,List[Measure]], List[Measure]] ) -> None:
        """Provides for a list of measures as modeled by the isd documentation. 
        Creates an intermediate level schema and allows applications to iterate over Measures, filling 
        the values from the record-string.

        Args:
            name (str): The section name.
            measures (Union[Callable[...,List[Measure]], List[Measure]]): Either a list of measure instances 
                or a callable that returns one.
        """
        self._name = name
        if callable(measures):
            measures = measures()    
        self._measures = measures 
    
    @property 
    def name(self) -> str: 
        return self._name 
    
    def measures(self) -> Generator[Measure, None, None]:
        """Yields measure instances,

        Yields:
            Generator[Measure]: Yields measure instances.
        """
        for m in self._measures:
            yield m 
            
    def schema(self) -> Dict[str, Any]:
        """Returns a schema that in turn produces schemas for each measure.

        Returns:
            Dict[str, Union[str, List[Dict[str, Any]]]]: A section schema of 
                name: name, measures: List[Measures]
        """
        return {
            'name': self._name, 
            'measures': [m.schema() for m in self._measures]
        }    
    


def control_measures() -> List[Measure]:
    """A convenience factory method that returns all of the ISD measures for the control section.

    Returns:
        List[Measure]: A List of measures from the isd control section.
    """
    return [
        Measure('usaf', Position(5,10)), 
        Measure('wban', Position(11,15)), 
        Measure('date', Position(16,23)), 
        Measure('time', Position(24,27)),
        CategoricalMeasure('data_source_flag', Position(28,28),missing='9', mapping=mappings.DATA_SOURCE_FLAG),
        NumericMeasure('latitude',Position(29,34), missing='+999999', unit='angular_degrees', scaling_factor=1000), 
        NumericMeasure('longitude',Position(35,41), missing='+999999', unit='angular_degrees', scaling_factor=1000), 
        CategoricalMeasure('code', Position(42,46), missing='99999', mapping=mappings.GEOPHYSICAL_REPORT_CODES), 
        NumericMeasure('elevation_dimension', Position(47,51), missing='+9999', scaling_factor=1, unit='meters'),
        Measure('call_letter_identifier', Position(52,56), missing='99999'), 
        CategoricalMeasure('quality_control_process_name', Position(57,60), mapping=mappings.QUALITY_CONTROL_PROCESS)
    ]



def mandatory_measures() -> List[Measure]:
    """A convenience factory method that returns all of the ISD measures for the mandatory section.

    Returns:
        List[Measure]: A List of measures from the isd mandatory section
    """
    return [
        NumericMeasure('wind_observation_direction_angle', Position(61,63), missing='999', scaling_factor=1, unit='angular_degrees'), 
        CategoricalMeasure('wind_observation_direction_quality_code', Position(64,64), mapping=mappings.NUMERICS),
        CategoricalMeasure('wind_observation_type_code', Position(65,65), missing='9', mapping=mappings.WIND_OBS_TYPE_CODE),
        NumericMeasure('wind_observation_speed_rate', Position(66,69), unit='meters_per_second', scaling_factor=10, missing='9999'), 
        CategoricalMeasure('wind_observation_speed_quality_code',Position(70,70), mapping=mappings.NUMERICS),
        NumericMeasure('sky_condition_observation_ceiling_height_dimension', Position(71,75), unit='meters', scaling_factor=1, missing='99999'),
        CategoricalMeasure('sky_condition_observation_ceiling_quality_code', Position(76,76), mapping=mappings.NUMERICS),
        CategoricalMeasure('sky_condition_observation_ceiling_determination_code', Position(77,77), missing='9', mapping=mappings.CEILING_DETERMINATION_CODE),
        CategoricalMeasure('sky_condition_observation_cavok_code', Position(78,78), missing='9', mapping=mappings.CAVOK_CODE), 
        NumericMeasure('visibility_observation_distance_dimension',Position(79,84), unit='meters', missing='999999'),
        CategoricalMeasure('visibility_observation_distance_quality_code', Position(85,85), mapping=mappings.NUMERICS),
        CategoricalMeasure('visibility_observation_variability_code',Position(86,86), missing='9', mapping=mappings.VARIABLITY_CODE),
        CategoricalMeasure('visibility_observation_quality_variability_code', Position(87,87), mapping=mappings.NUMERICS), 
        NumericMeasure('air_temperature_observation_air_temperature', Position(88,92), scaling_factor=10,unit='degrees_celsius', missing='+9999'),
        CategoricalMeasure('air_temperature_observation_air_temperature_quality_code', Position(93,93), mapping=mappings.NUMERICS),
        NumericMeasure('air_temperature_observation_dew_point_temperature', Position(94,98), unit='degrees_celsius', scaling_factor=10,missing='+9999'),
        CategoricalMeasure('air_temperature_observation_dew_point_quality_code', Position(99,99), mapping=mappings.NUMERICS),
        NumericMeasure('atmospheric_pressure_observation_sea_level_pressure', Position(100,104), unit='hectopascals', missing='99999', scaling_factor=10),
        CategoricalMeasure('atmospheric_pressure_observation_sea_level_pressure_quality_code', Position(105,105), mapping=mappings.NUMERICS)
    ]